File: top-interview-questions-medium/Trees and Graphs/populating-next-right-pointers-in-each-node/116.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None


class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if not root: return
        queue = [root]
        while len(queue):
            size = len(queue)
            level = queue[:]

            for i in range(size):
                curr = queue.pop(0)
                curr.next = level[i + 1] if i + 1 < len(level) else None

                if curr.left: queue.append(curr.left)
                if curr.right: queue.append(curr.right)

File: top-interview-questions-medium/Trees and Graphs/populating-next-right-pointers-in-each-node/116.js
/**
 * Definition for binary tree with next pointer.
 * function TreeLinkNode(val) {
 *     this.val = val;
 *     this.left = this.right = this.next = null;
 * }
 */

/**
 * @param {TreeLinkNode} root
 * @return {void} Do not return anything, modify tree in-place instead.
 */
var connect = function(root) {
  if (!root) return
  const queue = [root]
  while (queue.length) {
    const size = queue.length
    const level = [...queue]

    for (let i = 0; i < size; i++) {
      const currentNode = queue.shift()
      currentNode.next = level[i + 1]
      if (currentNode.left) queue.push(currentNode.left)
      if (currentNode.right) queue.push(currentNode.right)
    }
  }
}
